modules/app/src/main/resources/lib/collaboration.js

```diff
@@ -1,33 +1,133 @@
+const eventLib = require('/lib/xp/event');
+
+const TTL_SECONDS = 2 * 60 * 1000;
+
+const SHARED_MAP = sharedMemoryLib.getMap('contentstudio.collaboration.contents');
+
 function required(params, name) {
     const value = params[name];
-    if (value === undefined) {
+    if (typeof value === 'undefined' || value === null) {
         throw `Parameter "${name}" is required`;
     }
-    return value;
+}
+
+function CollaboratorsModifier(params) {
+    this.collaboratorKey = `${params.sessionId}=${params.userKey}`;
+    this.contentId = params.contentId;
+    this.collaborators = [];
+
+    this.removeExpired = function (timeAt) {
+        const self = this;
+        const originalSize = this.collaborators.length;
+        this.collaborators = this.collaborators.filter(collaborator => {
+            let lastJoin = self.extractLastJoin(collaborator);
+            return timeAt - lastJoin < TTL_SECONDS;
+        });
+        return originalSize > this.collaborators.length;
+    };
+
+    this.removeCurrent = function () {
+        const originalSize = this.collaborators.length;
+        const collaboratorKey = this.collaboratorKey;
+        this.collaborators = this.collaborators.filter(collaborator => !collaborator.startsWith(collaboratorKey));
+        return originalSize > this.collaborators.length;
+    };
+
+    this.getUniqueCollaborators = function () {
+        let uniqueCollaborators = new Set();
+        this.collaborators.forEach(collaborator => uniqueCollaborators.add(collaborator));
+
+        let collaborators = [];
+        uniqueCollaborators.forEach(collaborator => collaborators.push(collaborator));
+        this.collaborators = collaborators;
+
+        return this.collaborators;
+    };
+
+    this.publishEvent = function () {
+        const self = this;
+        const userKeys = this.getUniqueCollaborators().map(collaboratorId => self.extractUserKey(collaboratorId));
+        const contentId = this.contentId;
+
+        eventLib.send({
+            type: 'edit.content.collaborators.update',
+            distributed: true,
+            data: {
+                contentId: contentId,
+                collaborators: userKeys
+            }
+        });
+    };
+
+    this.extractLastJoin = function (collaboratorId) {
+        let idParts = collaboratorId.split("=", -1);
+        return parseInt(idParts[2]);
+    };
+
+    this.extractUserKey = function (collaboratorId) {
+        return collaboratorId.split("=", -1)[1];
+    };
+
+    this.getStoredCollaborators = function (storedCollaborators, isJoin) {
+        if (typeof storedCollaborators === 'undefined' || storedCollaborators === null) {
+            return isJoin ? [] : null;
+        }
+        return storedCollaborators;
+    };
+}
+
+CollaboratorsModifier.prototype.modify = function (oldCollaborators, isJoin) {
+    this.collaborators = this.getStoredCollaborators(oldCollaborators, isJoin);
+
+    if (isJoin === false && this.collaborators === null) {
```

Comment:
Why not use `!isJoin`? isJoin is sent explicitly so it won't have undefined value. 
you can write like this:
```
if (!isJoin && !this.collaborators) {
            return null;
        } 
```

Reply:
done